libportal/portal.c

```diff
@@ -273,3 +280,200 @@ xdp_portal_new (void)
 {
   return g_object_new (XDP_TYPE_PORTAL, NULL);
 }
+
+/**
+ * xdp_portal_running_under_flatpak:
+ * @error: return location for a #GError pointer
+ *
+ * If you are trying to determine whether to use portals, consider using
+ * xdp_portal_running_under_sandbox() instead.
+ *
+ * Returns: %TRUE if the current process is running under a Flatpak sandbox, or
+ *   %FALSE if either an error was encountered or the process is running
+ *   unsandboxed
+ */
+gboolean
+xdp_portal_running_under_flatpak (GError **error)
+{
+  static _Thread_local gboolean decided = FALSE;
+  static _Thread_local gboolean under_flatpak = FALSE;
+  g_autofree char *root_path = NULL;
+  int root_fd = -1;
+  int info_fd = -1;
+
+  if (decided)
+    return under_flatpak;
+
+  /* Note: the rest of this function is copied from xdg-desktop-portal */
+  root_path = g_strdup_printf ("/proc/%u/root", getpid ());
```

Comment:
This complication seems to exist entirely for discovering if *other* processes are running under flatpak. Surely the task is quite a bit easier when checking in your own root and much of this can be simplified?

Reply:
Well, it doesn't seem impossible that the current process is running on a fuse root filesystem, which is the main source of the complexity. So I guess it depends whether we value avoiding complexity with a simple `g_file_test ("/.flatpak-info", G_FILE_TEST_EXISTS);` more than not setting the `GError` in that case.

Reply:
However experimentally `/proc/getpid()/root` is readable from a process running under toolbox, so maybe you're right and we can get rid of a lot of this.

Reply:
I changed this to have separate API depending on whether its your own pid or another pid.

Reply:
Now changed it to use the simpler `g_file_test` check